### DISCUSSION TITLE
Disjoint Set(Union Find)

### DISCUSSION THREAD

**Discussion by BLoHny** [subak-ba/cs-study#47] [자료 구조]
# Disjoint Set

### 정의
Disjoin-Set(혹은 Union-Find)은 상호 배타적으로 이루어진 집합을 효율적으로 표현하기 위해 만들어진이다. Disjoin Set에서 서로 다른 두 개의 집합을 병합하는 Union 연산과 집합의 원소가 어떠한 집합에 속해있는지 판단하는 Find 연산을 지원하기 때문에 Union-Find라는 이름이 붙게 되었다.
> 상호 베타적의 의미 ->  ex) a는 b에 속하고 c도 b에 속하지만 a와 c는 다르다

### 연산
- FIND 연산 -> 자신이 속한 집합의 최상단 노드 반환
- UNION 연산 -> 두개의 집합을 하나로 합치는 연산

### 원시적 형태
ArrayList를 통해 상호 베타적 집합을 표현하는 간단한 방법은 배열의 요소에 번호를 부여하는 것이다. 이것은 탐색할 요소가 속한 집합이 어느 집합인지 쉽게 알 수 있어 시간복잡도가 O(1)이 걸리지만 Union에서는 두개의 배열을 합쳐야 하므로 기존 인덱스 번호를 옮기는 연산이 필요하므로 O(N)이 걸리므로 비효율 적이다. 이런 문제를 트리를 통해 해결한다.

<br>

### 기본적 형태
Disjoint Set에서는 트리의 구조 자체가 의미를 무시하고 단 하나의 최상위 노드에만 관심을 가진다. Disjoint Set의 트리 구조에서 최상위 노드는 각 집합을 대표하는 대표자 역할을 맡게 된다. Disjoint Set을 트리로 표현하기 위해서는 먼저 ArrayList의 각 원소에 자신의 인덱스 값이 들어가 있는 초기 상태가 필요하다. 이 상태에서 각 원소에 들어가 있는 값은 각 원소의 부모를 의미한다.

#### 초기 형태 
<img src ="https://github.com/subak-ba/cs-study/assets/113517233/f7360387-7bc7-4d93-8596-ed1084d0d40c" width="50%"><br>
* 자신의 값을 가리키는 형태

#### 집합이 존재하는 형태
<img src ="https://github.com/subak-ba/cs-study/assets/113517233/b63fa789-ceb4-4105-9b21-2e535ac74310" width="50%"><br>
* union을 통해 집합이 합쳐진 경우

```java
T find(x) {
  // 찾고자 하는 요소의 부모가 자기 자신일시 그대로 반환
  if (x == parent[x]) {
     return x;
  }
  
  //자기 자신을 가리키지 않을시 재귀호출을 진행함
  return find(parent[x]);
}
```

### UNION 연산
![image](https://github.com/subak-ba/cs-study/assets/113517233/fe2b4c45-93e5-4689-91c7-4d05ebe8c2fc)

#### 과정
- find 연산을 통해 두 요소의 루트 노드를 찾음.
- 두 루트 노드중 하나를 다른 트리의 루트 노드 아래에 배치하여 두 트리를 합침.

```java
T union(a, b) {
  //요소의 부모를 찾는 과정
  aParent = find(a);
  bParent = find(b);
   
  // 임의로 연결
  parent[bParent] = aParent;
} 
```

#### 단점
- 선형트리가 생길 수 있음 -> O(N)이 걸리게 됨
<br>

### 개선

- #### 경로 압축
   ![image](https://github.com/subak-ba/cs-study/assets/113517233/46df49be-6911-4ec9-9986-76f0b5daeda2)

  - find 연산시 그 노드를 자신의 루트에 직접적으로 연결하게 하는 방법
  - 압축후 같은 노드에 대한 find연산은 O(1)이 걸리게 됨.

```java
T find(x) {
 // 요소의 부모가 요소일시 그대로 반환
 if (x == parent[x]) {
   return x;
 }

 // 자신이 속한 집합의 최상단 노드에 바로 연결
 parent[x] = find(parent[x]);
 return parent[x];
}
```

- #### Union By Rank
  ![image](https://github.com/subak-ba/cs-study/assets/113517233/1ee7cb4b-0d16-4922-9f5a-9636d144e1e2)
 
  - 항상 작은 트리를 큰 트리에 붙이는 방법
  - 트리의 크기를 기록하는 배열을 이용함

#### size
```java
int[] rank = new int[MAX];
T union(a, b) {

  // 집합의 부모를 가져오는 과정
  aParent = find(a);
  bParent = find(b);
   
  //집합의 부모의 트리 사이즈 비교후 a, b값을 swap하거나 그대로 진행
  if (rank[aParent] < rank[bParent]) {
    int t = aParent;
    aParent = bParent;
    bParent = t;
  }
  
  // 부모로 지정
  parent[bParent] = a;

  // 사이즈 증가
  rank[aParent] += rank[bParent];
} 
```

위와 같은 개편을 통하여 최악의 경우 O(N)인 구조를 O(log N)으로 향상 시킬 수 있다.

### 사용처
크루스칼 알고리즘에서 최단 신장 트리를 찾기 위해 사용됨.

### 문제
https://www.acmicpc.net/problem/1717

**Comment by onee-only** [comment]
코드에 주석을 달아서 더 이해하기 쉽도록 만들어 주실 수 있나요?

**Reply by BLoHny** [reply]
수정하였습니다. 애매한 코드는 삭제했습니다.

**Comment by onee-only** [comment]
문제 풀었습니다!
http://boj.kr/12ba78af34d94aac9ea868d8deb36f61
